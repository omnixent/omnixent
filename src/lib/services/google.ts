import { ServiceArgs } from '.';
import { normalizeServiceArgs } from '../utils';

const endpoint = 'https://www.google.com/complete/search';

export function formatURI(options: ServiceArgs): string {
  const { t, hl } = normalizeServiceArgs(options);
  return `${endpoint}?client=gws-wiz&q=${t}&hl=${hl}`;
}

export function extractBody(response: string) {
  const body = response.replace(/^window\.google\.ac\.h\(/, '').replace(/\)$/, '');

  return JSON.parse(body)
    ?.flat()
    ?.flat()
    ?.filter((el: any) => typeof el === 'string')
    ?.map((term: string) => term.replace(/<b>/gi, ''))
    ?.map((term: string) => term.replace(/<\/b>/gi, ''))
    ?.map((term: string) => term.replace(/&#39;/gi, "'"));
}
