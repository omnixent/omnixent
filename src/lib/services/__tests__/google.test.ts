import { formatURI, extractBody } from '../google';

test('Testing google formatURI', () => {
  expect(formatURI({ term: 'Node.js', language: 'it', country: 'us' })).toMatchInlineSnapshot(
    `"https://www.google.com/complete/search?client=gws-wiz&q=node.js&hl=us-IT"`,
  );
  expect(
    formatURI({ term: 'blue jeans brands', language: 'en', country: 'de' }),
  ).toMatchInlineSnapshot(
    `"https://www.google.com/complete/search?client=gws-wiz&q=blue%20jeans%20brands&hl=de-EN"`,
  );
});

test('Testing google extractBody', () => {
  expect(
    extractBody(
      `window.google.ac.h([[["blue jeans brands",0,[19]],["blue jeans brands\u003cb\u003e list\u003c\/b\u003e",0,[19]],["blue jeans brands\u003cb\u003e from the 90s\u003c\/b\u003e",0,[19]],["\u003cb\u003etop \u003c\/b\u003eblue jeans brands",0,[22,30,19]],["\u003cb\u003eexpensive \u003c\/b\u003eblue jeans brands",0,[22,30,19]],["\u003cb\u003etarget \u003c\/b\u003eblue jeans brands",0,[22,30,19]],["\u003cb\u003emens \u003c\/b\u003eblue jeans brands",0,[22,30,19]],["\u003cb\u003ewestern \u003c\/b\u003eblue jeans brands",0,[22,30,19]],["\u003cb\u003efamous \u003c\/b\u003eblue jeans brands",0,[22,30,19]],["\u003cb\u003esargent \u003c\/b\u003eblue jeans brands",0,[22,30,19]]],{"i":"blue jeans brands","l":"1","q":"ByZYgqmh8L85teUO48aZdL7g8xE"}])`,
    ),
  ).toMatchSnapshot();
  expect(
    extractBody(
      `window.google.ac.h([[["node\u003cb\u003e.\u003c\/b\u003ejs",0],["node\u003cb\u003e.\u003c\/b\u003ejs\u003cb\u003e download\u003c\/b\u003e",0],["node\u003cb\u003e.\u003c\/b\u003ejs\u003cb\u003e server-side javascript\u003c\/b\u003e",0],["node\u003cb\u003e.\u003c\/b\u003ejs\u003cb\u003e tutorial\u003c\/b\u003e",0],["node\u003cb\u003e.\u003c\/b\u003ejs\u003cb\u003e what is\u003c\/b\u003e",0],["node\u003cb\u003e.\u003c\/b\u003ejs\u003cb\u003e express\u003c\/b\u003e",0],["node\u003cb\u003e.\u003c\/b\u003ejs\u003cb\u003e cos\u0026#39;Ã¨\u003c\/b\u003e",0],["node\u003cb\u003e.\u003c\/b\u003ejs\u003cb\u003e server-side javascript virus\u003c\/b\u003e",0],["node\u003cb\u003e.\u003c\/b\u003ejs\u003cb\u003e windows 7\u003c\/b\u003e",0],["node\u003cb\u003e.\u003c\/b\u003ejs\u003cb\u003e require\u003c\/b\u003e",0]],{"q":"ByZYgqmh8L85teUO48aZdL7g8xE"}])`,
    ),
  ).toMatchSnapshot();
});
