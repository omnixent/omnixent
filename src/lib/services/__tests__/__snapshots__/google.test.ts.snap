// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing google extractBody 1`] = `
Array [
  "blue jeans brands",
  "blue jeans brands list",
  "blue jeans brands from the 90s",
  "top blue jeans brands",
  "expensive blue jeans brands",
  "target blue jeans brands",
  "mens blue jeans brands",
  "western blue jeans brands",
  "famous blue jeans brands",
  "sargent blue jeans brands",
]
`;

exports[`Testing google extractBody 2`] = `
Array [
  "node.js",
  "node.js download",
  "node.js server-side javascript",
  "node.js tutorial",
  "node.js what is",
  "node.js express",
  "node.js cos'Ã¨",
  "node.js server-side javascript virus",
  "node.js windows 7",
  "node.js require",
]
`;
